### PR TITLE
Run sanitizers on tests

### PR DESCRIPTION
**Pull request by jacob-hughes**
This PR ensures that thread sanitizer and valgrind are run on `gc_tests` as part of the buildbot configuration now. 

After the initial setup, both valgrind and TSAN found quite a few bugs! So there are also fixes for each one.

### CODE REVIEW

File: src/alloc.rs [hunk 1]
```diff
@@ -168,7 +168,7 @@ impl<'a> Iterator for AllocListIterMut<'a> {
         // It's UB to call `.add` on a pointer past its allocation bounds, so we
         // need to check that it's within range before turning it into a pointer
         // and dereferencing it
-        if self.idx * core::mem::size_of::<Block>() >= SIZE_ALLOC_INFO {
+        if self.idx * size_of::<Block>() >= (SIZE_ALLOC_INFO - size_of::<Block>()) {
```
Comment (ltratt) [inline src/alloc.rs]: This change seems unrelated to this commit?

Comment (jacob-hughes) [reply src/alloc.rs]: Oops, that shouldn't have been folded into the previous commit. I can fix this when I squash

File: gc_tests/run_tests.rs [hunk 2]
```diff
@@ -65,6 +75,10 @@ fn main() {
             let lib_arg = [CRATE_NAME, "=", lib_path.to_str().unwrap()].concat();
             let deps_arg = ["dependency=", deps_path.to_str().unwrap()].concat();
 
+            let san_flags = sanitizers.iter().map(|x| format!("-Zsanitizer={}", x));
+
+            compiler.args(san_flags);
```
Comment (ltratt) [inline gc_tests/run_tests.rs]: I'm not sure the blank line between the assignment and (only) use is helpful.

Comment (jacob-hughes) [reply gc_tests/run_tests.rs]: Fixed: 89252e717bd6c9a5f8038ccd88df42b2017e241e

**Comment by ltratt** [comment]
Please squash and split.

**Comment by jacob-hughes** [comment]
Squashed.
